# Data Transformation

In [2]:
# import necessary packages

import pandas as pd
import numpy as np

In [3]:
# read google play store dataset

df = pd.read_csv('dataset/googleplaystore.csv')

In [4]:
df.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


Columns;

* **App**: Includes application name
* **Category**: Includes category of the application
* **Rating**: Includes the rate of application between 0 and 5
* **Review**: Includes the number of reviews that the app has
* **Size**: Size of the application (KB)
* **Installs**: Includes the number of user downloads/installs for the app
* **Type**: Includes app is free or paid 
* **Price**: Includes the price of the apps
* **Content Rating**: Includes age group the app is targeted at (Children / Mature 21+ / Adults, etc.)
* **Genres**: Includes the multiple genres the app belongs to (apart from its main category), these are separated by ';'
* **Last Updated**: Includes date of the app is updated lastly
* **Current Ver**: Includes current version of the app
* **Android Ver**: Includes android version of the app

In [6]:
df.describe()

Rating
count  9367.000000
mean      4.193338
std       0.537431
min       1.000000
25%       4.000000
50%       4.300000
75%       4.500000
max      19.000000

In [7]:
df.describe(include='object')

App Category Reviews                Size    Installs   Type  Price  \
count    10841    10841   10841               10841       10841  10840  10841   
unique    9660       34    6002                 462          22      3     93   
top     ROBLOX   FAMILY       0  Varies with device  1,000,000+   Free      0   
freq         9     1972     596                1695        1579  10039  10040   

       Content Rating Genres    Last Updated         Current Ver Android Ver  
count           10840  10841           10841               10833       10838  
unique              6    120            1378                2832          33  
top          Everyone  Tools  August 3, 2018  Varies with device  4.1 and up  
freq             8714    842             326                1459        2451

# Data Transformation

### Data Type

According to the attributes of the dataset, some numeric columns are available as an object type in the initial state. So, data transformation is necessary to analyze the dataset with all numeric and categorical features. Columns will be converted to numeric values in the following section;
* Reviews
* Size

In [8]:
# Reviews
# review column includes '3.0M'

df.loc[df['Reviews'] == '3.0M', 'Reviews'] = 3000000

# change Reviews column type to numeric value

df['Reviews'] = df['Reviews'].astype('Float64')

In [9]:
# Size
# convert size column

df.loc[df['Size']=='Varies with device', 'Size'] = '11M' # convert to the most common size
df.loc[df['Size']=='1,000+', 'Size'] = '11M' # convert to the most common size

# convert all MB values to KB
# convert all values to KB

for index, row in df.iterrows():
    # Access the value in the 'Reviews' column for the current row
    review_value = row['Size']

    # Check if the value contains 'M' (MB) and convert it to a numeric value
    if 'M' in review_value:
        numeric_value = float(review_value.replace('M', '')) * 1024
        df.at[index, 'Size'] = numeric_value
    elif 'k' in review_value:
        numeric_value = float(review_value.replace('k', ''))
        df.at[index, 'Size'] = numeric_value
    else: 
        print(review_value)

# change Size column type to float

df['Size'] = df['Size'].astype('Float64')
#df['Size'] = df['Size']/1024

In [10]:
# Installs

for index, row in df.iterrows():
    # Access the value in the 'Installs' column for the current row
    review_value = row['Installs']

    # Check if the value contains '+' and ',' and convert it to a numeric value
    if '+' in review_value:
        numeric_value = float(review_value.replace('+', '').replace(',', ''))
        df.at[index, 'Installs'] = numeric_value
    else: 
        df.at[index, 'Installs'] = 0

df['Installs'] = df['Installs'].astype('Float64')

In [11]:
# Price 

for index, row in df.iterrows():
    # Access the value in the 'Price' column for the current row
    review_value = row['Price']

    # Check if the value contains '$' and ',' and convert it to a numeric value
    if '$' in review_value:
        numeric_value = float(review_value.replace('$', ''))
        df.at[index, 'Price'] = numeric_value
    else: 
        df.at[index, 'Price'] = 0

df['Price'] = df['Price'].astype('Float64')

In [12]:
# convert '0' to categoric value

df.loc[df['Type']=='0', 'Type'] = 'Free'

In [13]:
# convert 'Date' column to datetime type with error handling
# convert cell with errors to NaN

df['Last Updated'] = pd.to_datetime(df['Last Updated'], errors='coerce')

After conversions, the description of the numeric and categorical variables can be checked again to see the current version.

In [14]:
df.describe()

Rating         Reviews          Size         Installs      Price
count  9367.000000         10841.0       10841.0          10841.0    10841.0
mean      4.193338   444388.653537  20347.887178   15462912.41463   1.027273
std       0.537431  2927728.464012  21601.967928  85025569.008648  15.948971
min       1.000000             0.0           8.5              0.0        0.0
25%       4.000000            38.0        6041.6           1000.0        0.0
50%       4.300000          2094.0       11264.0         100000.0        0.0
75%       4.500000         54798.0       26624.0        5000000.0        0.0
max      19.000000      78158306.0      102400.0     1000000000.0      400.0

In [15]:
df.describe(include='object')

App Category   Type Content Rating Genres         Current Ver  \
count    10841    10841  10840          10840  10841               10833   
unique    9660       34      2              6    120                2832   
top     ROBLOX   FAMILY   Free       Everyone  Tools  Varies with device   
freq         9     1972  10040           8714    842                1459   

       Android Ver  
count        10838  
unique          33  
top     4.1 and up  
freq          2451

### Missing Values

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype         
---  ------          --------------  -----         
 0   App             10841 non-null  object        
 1   Category        10841 non-null  object        
 2   Rating          9367 non-null   float64       
 3   Reviews         10841 non-null  Float64       
 4   Size            10841 non-null  Float64       
 5   Installs        10841 non-null  Float64       
 6   Type            10840 non-null  object        
 7   Price           10841 non-null  Float64       
 8   Content Rating  10840 non-null  object        
 9   Genres          10841 non-null  object        
 10  Last Updated    10840 non-null  datetime64[ns]
 11  Current Ver     10833 non-null  object        
 12  Android Ver     10838 non-null  object        
dtypes: Float64(4), datetime64[ns](1), float64(1), object(7)
memory usage: 1.1+ MB


In [17]:
# count missing values in a column

for col_name in df.columns:
    count_nan = df[col_name].isna().sum()
    print ('Column ' + col_name + ": ", count_nan)

Column App:  0
Column Category:  0
Column Rating:  1474
Column Reviews:  0
Column Size:  0
Column Installs:  0
Column Type:  1
Column Price:  0
Column Content Rating:  1
Column Genres:  0
Column Last Updated:  1
Column Current Ver:  8
Column Android Ver:  3


As you can see above, some numeric and categorical columns include Null values. To deal with Null values;
1. These rows can be removed from the dataset
2. These Null cells can be filled with values belonging to the columns

Generally although removing the rows is preferred, sometimes there can be lots of Null cells inside the data set, and removing one row can cause to loss of other information through other columns. So, for this study filling the missing values with the proper descriptive values of columns is preferred. 

* For the Review column, missing cells will be filled with the mean of all rating
* For the Type column, one missing cell will be filled with the most frequent type.
* For the Content Rating column, one missing cell will be filled with the most frequent type.

Other columns with missing values (Last Updated, Current Ver, Android Ver) will be dealt with in case of need.

In [18]:
# fill with mean of rating

df['Rating'] = np.where((df['Rating'].isnull()), df['Rating'].mean(), df['Rating'])

# fill with frequent values

df['Type'] = np.where((df['Type'].isnull()), 'Free', df['Type'])
df['Content Rating'] = np.where((df['Content Rating'].isnull()), 'Everyone', df['Content Rating'])

On the other hand, if you want to decide to drop the columns which have more than the determined threshold missing value rate. You can remove it by running the following code.

In [19]:
# drop the columns which have more than %50 missing value rate

for col_name in df.columns:
    perc_nan = df[col_name].isna().sum()/len(df)*100
    if perc_nan >= 50:
        print(col_name)
        df.drop(columns={col_name}, inplace=True)

If you want to sort the dataframe relative to one column;

In [20]:
df_sorted = df.sort_values('Installs', ascending = False)
df_sorted.head()

App       Category  Rating  \
3896                            Subway Surfers           GAME     4.5   
3943                                  Facebook         SOCIAL     4.1   
335   Messenger – Text and Video Chat for Free  COMMUNICATION     4.0   
3523                              Google Drive   PRODUCTIVITY     4.4   
3565                              Google Drive   PRODUCTIVITY     4.4   

         Reviews     Size      Installs  Type  Price Content Rating  \
3896  27711703.0  77824.0  1000000000.0  Free    0.0   Everyone 10+   
3943  78128208.0  11264.0  1000000000.0  Free    0.0           Teen   
335   56642847.0  11264.0  1000000000.0  Free    0.0       Everyone   
3523   2731211.0  11264.0  1000000000.0  Free    0.0       Everyone   
3565   2731211.0  11264.0  1000000000.0  Free    0.0       Everyone   

             Genres Last Updated         Current Ver         Android Ver  
3896         Arcade   2018-07-12              1.90.0          4.1 and up  
3943         Social   2018-08-03  Varies with device  Varies with device  
335   Communication   2018-08-01  Varies with device  Varies with device  
3523   Productivity   2018-08-06  Varies with device  Varies with device  
3565   Productivity   2018-08-06  Varies with device  Varies with device

Also, one outlier is detected in the Rating column. The rating values should be between 0 and 5.

In [21]:
df[df['Rating']>5]

App Category  Rating    Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0  3000000.0   

          Size  Installs  Type  Price Content Rating             Genres  \
10472  11264.0       0.0  Free    0.0       Everyone  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472          NaT  4.0 and up         NaN

So, this value is removed from dataset.

In [22]:
df = df[df['Rating']<=5]

In [23]:
df.describe(include='object')

App Category   Type Content Rating Genres         Current Ver  \
count    10840    10840  10840          10840  10840               10832   
unique    9659       33      2              6    119                2831   
top     ROBLOX   FAMILY   Free       Everyone  Tools  Varies with device   
freq         9     1972  10040           8714    842                1459   

       Android Ver  
count        10838  
unique          33  
top     4.1 and up  
freq          2451

It's interesting the app name has less number of unique values relative to the dataset row count. So Let's check if there are any duplicated rows. But of course, there can be an app that has the same name with another name. Checking the duplicated values inside the dataset is another data preparation process that should not be missed.


In [25]:
duplicates = df[df.duplicated()]

duplicates

App      Category  Rating  \
229                       Quick PDF Scanner + OCR FREE      BUSINESS     4.2   
236                                                Box      BUSINESS     4.2   
239                                 Google My Business      BUSINESS     4.4   
256                                ZOOM Cloud Meetings      BUSINESS     4.4   
261                          join.me - Simple Meetings      BUSINESS     4.0   
...                                                ...           ...     ...   
8643                    Wunderlist: To-Do List & Tasks  PRODUCTIVITY     4.6   
8654   TickTick: To Do List with Reminder, Day Planner  PRODUCTIVITY     4.6   
8658                           ColorNote Notepad Notes  PRODUCTIVITY     4.6   
10049        Airway Ex - Intubate. Anesthetize. Train.       MEDICAL     4.3   
10768                                             AAFP       MEDICAL     3.8   

         Reviews     Size     Installs  Type  Price Content Rating  \
229      80805.0  11264.0    5000000.0  Free    0.0       Everyone   
236     159872.0  11264.0   10000000.0  Free    0.0       Everyone   
239      70991.0  11264.0    5000000.0  Free    0.0       Everyone   
256      31614.0  37888.0   10000000.0  Free    0.0       Everyone   
261       6989.0  11264.0    1000000.0  Free    0.0       Everyone   
...          ...      ...          ...   ...    ...            ...   
8643    404610.0  11264.0   10000000.0  Free    0.0       Everyone   
8654     25370.0  11264.0    1000000.0  Free    0.0       Everyone   
8658   2401017.0  11264.0  100000000.0  Free    0.0       Everyone   
10049      123.0  88064.0      10000.0  Free    0.0       Everyone   
10768       63.0  24576.0      10000.0  Free    0.0       Everyone   

             Genres Last Updated         Current Ver         Android Ver  
229        Business   2018-02-26  Varies with device        4.0.3 and up  
236        Business   2018-07-31  Varies with device  Varies with device  
239        Business   2018-07-24    2.19.0.204537701          4.4 and up  
256        Business   2018-07-20      4.1.28165.0716          4.0 and up  
261        Business   2018-07-16           4.3.0.508          4.4 and up  
...             ...          ...                 ...                 ...  
8643   Productivity   2018-04-06  Varies with device  Varies with device  
8654   Productivity   2018-08-06  Varies with device  Varies with device  
8658   Productivity   2018-06-27  Varies with device  Varies with device  
10049       Medical   2018-06-01              0.6.88          5.0 and up  
10768       Medical   2018-06-22               2.3.1          5.0 and up  

[483 rows x 13 columns]

So,  as you can see above, the dataset has 483 duplicated rows. They should be removed from the dataset not to cause misleading.

In [26]:
df.drop_duplicates(inplace=True)

In [28]:
# save transformed dataset

df.to_csv('dataset/transformed-googleplaystore.csv', index=False)